core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -400,13 +392,14 @@ private[zio] final class FiberContext[E, A](
 
                   if (stack.isEmpty) {
                     // Error not caught, stack is empty:
-                    curZio = null
+                    val cause = {
+                      // Add interruption information into the cause, if it's not already there:
+                      val interrupted = state.get.interrupted
 
-                    val cause =
-                      if (interrupted && !cause0.interrupted) cause0 ++ Cause.interrupt
-                      else cause0
+                      if (!cause0.contains(interrupted)) cause0 ++ interrupted else cause0
```

Comment:
That contains call could get pretty expensive for large causes. I don't have a good idea on how to avoid it, but it's only relevant for internal interruption, right?

Reply:
Even external interruption. There must be a better way to avoid it. I got closer this time to eliminating the hackiness but there's still a few small issues that have to be dealt with.